pulumi-programs/ttl-stacks/index.ts

```diff
@@ -1,251 +1,254 @@
-
-import * as awsx from "@pulumi/awsx";
 import * as aws from "@pulumi/aws";
+import * as awsx from "@pulumi/awsx";
 import * as pulumi from "@pulumi/pulumi";
-import fetch from "node-fetch";
-
+import * as service from "@pulumi/pulumiservice";
+import * as random from "@pulumi/random";
 import * as crypto from "crypto";
+import fetch from "node-fetch";
 
 const config = new pulumi.Config();
+const region = aws.config.requireRegion();
 
-const stackConfig = {
-    // Webhook secret used to authenticate messages. Must match the value on the
-    // webhook's settings.
-    sharedSecret: config.get("sharedSecret"),
-    pulumiAccessToken: config.requireSecret("pulumiAccessToken"),
-};
+// Webhook secret used to authenticate messages. Must match the value on the
+// webhook's settings.
+const sharedSecret = new random.RandomString("shared-secret", { length: 16 });
 
-// Just logs information from an incoming webhook request.
-function logRequest(req: awsx.apigateway.Request) {
-    const webhookID = req.headers !== undefined ? req.headers["pulumi-webhook-id"] : "";
-    const webhookKind = req.headers !== undefined ? req.headers["pulumi-webhook-kind"] : "";
-    console.log(`Received webhook from Pulumi ${webhookID} [${webhookKind}]`);
-}
+// The Pulumi token our Destroy lambda will use with Automation API.
+const pulumiAccessToken = config.requireSecret("pulumiAccessToken");
 
-// Webhooks can optionally be configured with a shared secret, so that webhook handlers like this app can authenticate
-// message integrity. Rejects any incoming requests that don't have a valid "pulumi-webhook-signature" header.
-function authenticateRequest(req: awsx.apigateway.Request): awsx.apigateway.Response | undefined {
-    const webhookSig = req.headers !== undefined ? req.headers["pulumi-webhook-signature"] : "";
-    if (!stackConfig.sharedSecret || !webhookSig) {
-        return undefined;
-    }
+// We'll run our Destroy lambda in a container to get our package dependencies
+// set up.
+const image = awsx.ecr.buildAndPushImage("stack-ttl", {
+  context: "./app",
+});
+
+// Permissions for our lambdas and step functions. These are appropriate for a
+// demo but too broad for production.
+const lambdaRole = new aws.iam.Role("stack-ttl-lambda-role", {
+  assumeRolePolicy: aws.iam.assumeRolePolicyForPrincipal({
+    Service: "lambda.amazonaws.com",
+  }),
+  managedPolicyArns: [
+    aws.iam.ManagedPolicy.AWSLambdaBasicExecutionRole,
+    aws.iam.ManagedPolicy.AmazonS3FullAccess,
+    aws.iam.ManagedPolicy.AWSStepFunctionsFullAccess,
```

Comment:
I believe we basically want this lambda to have full aws list and delete access. That way the lambda pass along the session-scoped credentials at runtime to the deployment.